File: windows/src/RealtimeTranslator.Core/Realtime/TranslationTargetSelector.cs
using RealtimeTranslator.Core.Audio;
using RealtimeTranslator.Core.OpenAI;

namespace RealtimeTranslator.Core.Realtime;

public readonly record struct TranslationTargetSelection(
    RealtimeTranslationOutputLanguage? Target,
    int ReverseEvidenceCount);

/// <summary>pair と evidence から翻訳出力 target を調停する純粋な状態遷移。</summary>
public static class TranslationTargetSelector
{
    public const int ScriptSwitchMinimumLatinWords = 4;
    public const int ScriptSwitchMinimumLatinScalars = SpokenLanguageDetector.RecentEvidenceWindow;
    public const int ScriptSwitchMinimumJapaneseScalars = 2;

    public static TranslationTargetSelection Select(
        LanguagePair pair,
        RealtimeTranslationOutputLanguage? currentTarget,
        int reverseEvidenceCount,
        SpokenLanguageEvidence evidence,
        OppositeScriptRun? oppositeRun = null)
    {
        var candidate = CandidateTarget(pair, evidence, currentTarget is null);
        if (candidate is null)
        {
            return new(currentTarget, 0);
        }

        if (currentTarget == candidate)
        {
            return new(currentTarget, 0);
        }

        if (currentTarget is null)
        {
            return new(candidate, 0);
        }

        if (pair != LanguagePair.EnEs)
        {
            var shouldSwitch = oppositeRun is { } run
                && evidence switch
                {
                    SpokenLanguageEvidence.English or SpokenLanguageEvidence.Spanish =>
                        run.LatinWordCount >= ScriptSwitchMinimumLatinWords
                        && run.LatinScalarCount >= ScriptSwitchMinimumLatinScalars,
                    SpokenLanguageEvidence.Japanese =>
                        run.JapaneseScalarCount >= ScriptSwitchMinimumJapaneseScalars,
                    _ => false,
                };
            return new(shouldSwitch ? candidate : currentTarget, 0);
        }

        var nextCount = reverseEvidenceCount + 1;
        return nextCount >= 2
            ? new(candidate, 0)
            : new(currentTarget, nextCount);
    }

    private static RealtimeTranslationOutputLanguage? CandidateTarget(
        LanguagePair pair,
        SpokenLanguageEvidence evidence,
        bool isInitial)
    {
        if (evidence == SpokenLanguageEvidence.AmbiguousLatin
            && isInitial
            && pair != LanguagePair.EnEs)
        {
            var latinLanguage = pair.Counterpart(SpokenLanguage.Japanese);
            return latinLanguage is { } language
                ? pair.TranslationTarget(language)
                : null;
        }

        return evidence switch
        {
            SpokenLanguageEvidence.Japanese => pair.TranslationTarget(SpokenLanguage.Japanese),
            SpokenLanguageEvidence.English => pair.TranslationTarget(SpokenLanguage.English),
            SpokenLanguageEvidence.Spanish => pair.TranslationTarget(SpokenLanguage.Spanish),
            _ => null,
        };
    }
}

File: RealtimeTranslator/Realtime/TranslationTargetSelector.swift
import Foundation

struct TranslationTargetSelection: Equatable, Sendable {
    let target: RealtimeTranslationOutputLanguage?
    let reverseEvidenceCount: Int
}

enum TranslationTargetSelector {
    static let scriptSwitchMinimumLatinWords = 4
    static let scriptSwitchMinimumLatinScalars = SpokenLanguageDetector.recentEvidenceWindow
    static let scriptSwitchMinimumJapaneseScalars = 2

    static func select(
        pair: LanguagePair,
        currentTarget: RealtimeTranslationOutputLanguage?,
        reverseEvidenceCount: Int,
        evidence: SpokenLanguageEvidence,
        oppositeRun: OppositeScriptRun? = nil
    ) -> TranslationTargetSelection {
        let initial = currentTarget == nil
        guard let candidate = candidateTarget(pair: pair, evidence: evidence, isInitial: initial)
        else {
            return TranslationTargetSelection(
                target: currentTarget,
                reverseEvidenceCount: 0
            )
        }

        guard let currentTarget else {
            return TranslationTargetSelection(target: candidate, reverseEvidenceCount: 0)
        }
        guard candidate != currentTarget else {
            return TranslationTargetSelection(target: currentTarget, reverseEvidenceCount: 0)
        }

        if pair != .enEs {
            guard let oppositeRun else {
                return TranslationTargetSelection(target: currentTarget, reverseEvidenceCount: 0)
            }
            let shouldSwitch: Bool
            switch evidence {
            case .english, .spanish:
                shouldSwitch =
                    oppositeRun.latinWordCount >= scriptSwitchMinimumLatinWords
                    && oppositeRun.latinScalarCount >= scriptSwitchMinimumLatinScalars
            case .japanese:
                shouldSwitch = oppositeRun.japaneseScalarCount >= scriptSwitchMinimumJapaneseScalars
            case .ambiguousLatin, .none:
                shouldSwitch = false
            }
            return TranslationTargetSelection(
                target: shouldSwitch ? candidate : currentTarget,
                reverseEvidenceCount: 0
            )
        }

        let nextCount = reverseEvidenceCount + 1
        return TranslationTargetSelection(
            target: nextCount >= 2 ? candidate : currentTarget,
            reverseEvidenceCount: nextCount >= 2 ? 0 : nextCount
        )
    }

    private static func candidateTarget(
        pair: LanguagePair,
        evidence: SpokenLanguageEvidence,
        isInitial: Bool
    ) -> RealtimeTranslationOutputLanguage? {
        if evidence == .ambiguousLatin, isInitial, pair != .enEs {
            guard let latinLanguage = pair.counterpart(of: SpokenLanguage.japanese) else {
                return nil
            }
            return pair.translationTarget(for: latinLanguage)
        }

        switch evidence {
        case .japanese:
            return pair.translationTarget(for: .japanese)
        case .english:
            return pair.translationTarget(for: .english)
        case .spanish:
            return pair.translationTarget(for: .spanish)
        case .ambiguousLatin, .none:
            return nil
        }
    }
}
